game-app/game-core/src/main/java/games/strategy/triplea/ai/pro/ProCombatMoveAi.java

```diff
@@ -1197,6 +1156,91 @@ && canAirSafelyLandAfterAttack(unit, t)) {
+    final boolean raidsMayBePrecededByAirBattles =
+        Properties.getRaidsMayBePreceededByAirBattles(data.getProperties());
+    territoryManager
+        .getAttackOptions()
+        .getBomberMoveMap()
+        .entrySet()
+        .forEach(
+            bomberEntry -> {
+              final Unit bomber = bomberEntry.getKey();
+              if (alreadyAttackedWithUnits.contains(bomber)) {
+                return; // already attacked bombers cannot move
+              }
+              Set<Territory> bomberTargetTerritories = bomberEntry.getValue();
+              if (raidsMayBePrecededByAirBattles) {
+                bomberTargetTerritories =
+                    new HashSet(
+                        CollectionUtils.getMatches(
+                            bomberTargetTerritories,
+                            terr ->
+                                AirBattle.territoryCouldPossiblyHaveAirBattleDefenders(
+                                    terr, player, data, true)));
+              }
+              determineBestBombingAttackForBomber(
+                  attackMap, sortedUnitAttackOptions, bomberTargetTerritories, bomber);
+            });
+  }
+
+  /**
+   * Determines the best bombing attack territory for a single {@code bomber}.
+   *
+   * @param attackMap Attack map
+   * @param sortedUnitAttackOptions Sorted attack options
+   * @param bomberTargetTerritories Target territories
+   * @param bomber Bomber unit
+   */
+  private void determineBestBombingAttackForBomber(
+      final Map<Territory, ProTerritory> attackMap,
+      final Map<Unit, Set<Territory>> sortedUnitAttackOptions,
+      final Set<Territory> bomberTargetTerritories,
+      final Unit bomber) {
+    final Predicate<Unit> bombingTargetMatch =
+        Matches.unitCanProduceUnitsAndCanBeDamaged()
+            .and(Matches.unitIsLegalBombingTargetBy(bomber));
+    Optional<Territory> maxBombingTerritory = Optional.empty();
+    int maxBombingScore = MIN_BOMBING_SCORE;
+    for (final Territory t : bomberTargetTerritories) {
+      final List<Unit> targetUnits = t.getUnitCollection().getMatches(bombingTargetMatch);
+      if (!targetUnits.isEmpty() && canAirSafelyLandAfterAttack(bomber, t)) {
+        final int noAaBombingDefense =
+            t.anyUnitsMatch(Matches.unitIsAaForBombingThisUnitOnly()) ? 0 : 1;
+        int maxDamageProduction = TerritoryAttachment.getProduction(t);
+        int neededDamageUnits = 0;
+        final List<Unit> existingAttackingBombers = attackMap.get(t).getBombers();
+        final List<Unit> sameTargetBombers = new ArrayList<>();
+        for (Unit target : targetUnits) {
+          neededDamageUnits += target.getHowMuchMoreDamageCanThisUnitTake(t);
+          for (Unit existingBomber : existingAttackingBombers) {
+            if (!sameTargetBombers.contains(existingBomber)
+                && Matches.unitIsLegalBombingTargetBy(existingBomber).test(target)) {
+              sameTargetBombers.add(existingBomber);
```

Comment:
Don't we want to only add if *all* targets correspond? Right now, you add if *any* correspond.

You can change to something like:
```java
for (Unit existingBomber : existingAttackingBombers) {
  if (targetUnits.stream().allMatch(u -> Matches.unitIsLegalBombingTargetBy(existingBomber).test(u))) {
    sameTargetBombers.add(existingBomber);
  }
}
```

And have a separate loop to sum up `neededDamageUnits`.

Reply:
In case a bomber cannot raid against all units, but against one (do not know if this is actually possible) this code should consider the bomber as co-bomber if _any_ common target unit exists. Why should the co-bombers be limited to those being able to bomb _all_ target units?

Considering `neededDamageUnits` the idea is that all target units can consume damage before the production is harmed.

Reply:
> @asvitkine Fixes made. If you have further points, maybe we should set up a call instead of an extended back-and-forth here.

I think this the tricky code around bombing requires offline studying to have meaningful discussion, so while I appreciate the offer, I don't think I could provide good feedback in a live discussion on this.

Reply:
Okay, then we'll leave it?

Reply:
> In case a bomber cannot raid against all units, but against one (do not know if this is actually possible)
I believe this is possible, since each unit type can specify `getBombingTargets()` separately.

> this code should consider the bomber as co-bomber if any common target unit exists. Why should the co-bombers be limited to those being able to bomb all target units?

I think it gets tricky because you don't know which units something is targeting. For example, you have A that can target X and Y and B that can target Z and X. Let's say X,Y,Z are all in the same territory.

When considering B after A is already scheduled, you don't know if A's damage should reduce B's potential or not. So if you just consider units that both can target, you miss out on the fact that A could bomb just Y and B still has full potential between X and Z.

However, I think the case where we consider "all targets possible" is also not right, so I think neither of these simple approaches handle the above example well. As such, I'm OK with your current code on this - it's not perfect, but for most maps it will work well enough.

In the future, the logic could be expanded to handle complicated cases well, if some map runs into it, but that doesn't seem to be a priority.

Maybe just add a comment about the limitation?